### ISSUE TITLE
train_on_responses_only: false still trains on a doubled BOS, now off by one token from inference after #782

### ISSUE THREAD

**Issue by MakazhanAlpamys** [MakazhanAlpamys/Soup#785]
Follow-up to #781 / #782. Analysis is @Konuktor's, from the "not covered" section of #781; this issue exists so that work is not left inside a closed issue.

## What #782 fixed

Every **in-process** inference path (`chat`, `serve` transformers + `--mole`, `infer` / `bench`, `diff`, `data generate` local, the `live_eval` generators) now encodes rendered chat-template text with `add_special_tokens=False`, matching Soup's default SFT path (`train_on_responses_only: true`) and `apply_chat_template(tokenize=True)`.

### 1. Training on the non-default text path still bakes in the doubled BOS

With `train_on_responses_only: false`, `_legacy_text_format_row` produces rendered text that TRL's `tokenize_fn` tokenizes via `processing_class(text=...)` — i.e. **with** the tokenizer's special tokens on top of the template's. So those adapters train on `[bos, bos, ...]`. Before #782 that matched inference by accident; after #782 inference sends one BOS, so these adapters are now **off by one token** at inference.

`soup data preprocess` (`commands/data.py::preprocess_dataset`) tokenizes the same way.

This was read from the trl 0.29.0 / 0.29.1 source (the supported range), not executed.

**Why it needs its own change:** fixing it changes training data and therefore the pipeline cache key, so previously cached datasets must not be silently reused.

### 2. vLLM / SGLang / MII are unverified

These engines take the prompt string and tokenize inside the engine, so #782 deliberately left `build_chat_prompt` unchanged for them. Whether they add a second BOS is unmeasured — nobody involved had the hardware.

## Proposed fix path

1. Encode the legacy text path's rendered template text with no tokenizer special tokens (pre-tokenize before TRL, or pass the equivalent option through), in both the SFT row builder and `soup data preprocess`; include the change in the preprocess cache key so stale caches are invalidated rather than reused.
2. For each in-engine backend, one request on a template that renders `{{ bos_token }}` and inspect `final_output.prompt_token_ids[:2]` (already read in `utils/vllm.py`). If doubled, pass the engine's equivalent of "no special tokens" or hand it token ids.

## Acceptance criteria

- [ ] A test proving the `train_on_responses_only: false` path trains on exactly one BOS for a template that renders `{{ bos_token }}`, and zero for Soup's `data.chat_template` presets — using a real fast tokenizer, as `tests/test_issue781_inference_bos.py` does.
- [ ] That test fails on current `main`.
- [ ] `soup data preprocess` cache key changes, with a test.
- [ ] A changelog fragment stating that adapters trained with `train_on_responses_only: false` before this change see one extra BOS in training compared with inference.
- [ ] vLLM / SGLang / MII: a measured `prompt_token_ids[:2]` for each, recorded in the PR, and a fix where doubled. (Can be split into a separate PR if hardware is the blocker.)

**Comment by abdulwaarith0** [comment]
I'd like to take this. Checked and there's no open PR or claim on it.

Part 1 (the training-side correctness bug) is what I'd deliver first:
- Stop the legacy text path (`_legacy_text_format_row` in `data/sft_format.py`) from double-adding BOS, so `train_on_responses_only: false` trains on one BOS, matching what #782 made inference send. Either pre-tokenize with `add_special_tokens=False` before TRL or pass the equivalent option through its tokenize step. Same change for `soup data preprocess` (`preprocess_dataset` in `commands/data.py`).
- Fold it into the preprocess cache key (`make_preprocess_cache_key`) so previously cached datasets are invalidated, not silently reused.
- Tests with a real fast tokenizer, in the shape of `test_issue781_inference_bos.py`: prove the false path trains on exactly one BOS for a template that renders `{{ bos_token }}` and zero for the `data.chat_template` presets, and that it fails on current main. Plus a test that the cache key changes, and a changelog fragment noting adapters trained with `train_on_responses_only: false` before this carry one extra BOS versus inference.

Part 2 (vLLM / SGLang / MII `prompt_token_ids[:2]` measurement) I'll attempt as well: the check only needs a tiny model that renders `{{ bos_token }}`, so GPU memory isn't the constraint. These engines are Linux-first, so I'll run them under WSL2 and record `prompt_token_ids[:2]` for each that stands up, fixing where a second BOS appears. I'll flag any engine I genuinely can't get running rather than guess, and split it out if it holds part 1 back.

I'll build part 1 against current main.

**Comment by MakazhanAlpamys** [comment]
@abdulwaarith0 — yours. The plan matches the issue, and splitting part 2 out if the engines hold part 1 back is exactly right. Three things that decide whether part 1 fixes the BOS without breaking something next to it:

1. **Do not lose the EOS in training.** On the inference side, dropping a tokenizer-appended trailing EOS was the desired effect of #782. On the training side it is the opposite: if a template does not render an EOS and the tokenizer was the one appending it, `add_special_tokens=False` silently removes the token the model learns to stop on, which teaches run-on generation — the failure `soup data doctor`'s `eos_in_labels` check exists to catch. Please add an assertion that the **trained EOS count per row is unchanged** by your fix, for both a template that renders EOS and one that does not, and state in the PR what happens in each case.
2. **Apply the rule only to text a chat template rendered**, mirroring #782's `templated` flag. Rows that never went through a template (plain text, pretrain-style data, a model with no chat template) should keep the tokenizer's special tokens, because there the tokenizer's BOS is the only one. A no-template control that stays byte-identical to `main` is the test for that.
3. **Pre-tokenizing before TRL changes what TRL receives** (ids instead of `text`). Show that truncation to `max_length`, packing, and the resulting labels are otherwise identical to `main` — i.e. that the only diff in the token stream is the removed duplicate BOS. A before/after comparison on a handful of real rows in the PR body would settle it.

The cache-key bump and the changelog fragment wording in your plan are right as written.

One scheduling question, since #745 (LoRA+ in the preference and RL trainers) is also assigned to you: which one are you doing first? Either order is fine — I would just like the thread to say, so neither looks stalled.

**Comment by abdulwaarith0** [comment]
Doing #785 first since it's already in flight, then #745. Thanks for the three points, the EOS one especially. You're right that a blanket add_special_tokens=False drops a tokenizer-appended EOS when the template renders none, so I'll preserve the training EOS and add the eos_in_labels-style assertion for both a template that renders EOS and one that does not, plus the no-template byte-identical control and a before/after on real rows.

**Comment by MakazhanAlpamys** [comment]
Closed by #788 (merged as 1c399f1) by @abdulwaarith0, with the serving half in #867 (c083144) and the dependency floor it implies in #892 (6b42cc3). All three ship in the next release.

**Training side (#788).** The SFT path pre-tokenizes the rendered chat template itself, so TRL's `tokenize_fn` never runs and cannot prepend a second BOS. Because that also bypasses TRL's `add_eos`, `append_training_eos` reproduces TRL's rule directly — append `eos_token` when the rendered text does not already end with it — rather than probing the tokenizer's post-processor. Measured on real fast tokenizers across 12 template x post-processor shapes, and again across 24 `soup data preprocess` cache cases including truncated rows: **36 of 36 keep the stop token**, BOS goes 2 -> 1 where the template renders one and stays 0 for presets that render none.

**Serving side (#867).** `build_engine_prompt` encodes in-process and hands vLLM token ids instead of a string it would re-tokenize. Note the opposite case, which is a real change: a model whose template renders no BOS but whose post-processor would add one now gets **zero** at the engine rather than one.

**What was deliberately left out**, each with its own issue rather than folded in:

- **#890** — `soup serve --backend sglang` has the identical defect, with the measured `[128000, 128000]`.
- **#891** — same for `--backend mii`.
- **#876** — `soup data preprocess` cache and the live path still disagree on the leading BOS for `chat_template` presets. `preprocess` deliberately keeps main's tokenization here so its EOS stays byte-identical.

Rows that previously trained on two trailing EOS tokens now train on one; that is in the changelog.

Try it:

```
soup data doctor train.jsonl --model <your-base> --show-mask 3
```

The rendered rows now show a single leading BOS, and `--show-mask` goes through the same label builder the collator uses rather than a mirror of it — so what you see is what trains.
